### PR TITLE
Add test for starting kubelet with a CSI volume mounted

### PR DESCRIPTION
**Pull request by jsafrane**
#### What type of PR is this?
/kind feature

#### What this PR does / why we need it:
To have a regression tests for https://github.com/kubernetes/kubernetes/issues/117745, i.e. restart kubelet with a CSI volume mounted *and* the API server running as a static pod. We expect that the API server (or any other static pod) starts.

The test heavily uses `kind` containers and the fact that it runs the API server as a static pod.

I am adding it to sig-storage tests, I won't object if someone adds more similar tests and moves it to sig-node.

#### Which issue(s) this PR fixes:

Related to #117745

#### Special notes for your reviewer:

High level plan:

1. Add this e2e test
2. Create new periodic + optional pull CI jobs that runs a kind cluster + this e2e test (`--focus=[Feature:Kind]`): https://github.com/kubernetes/test-infra/pull/29544
3. Create a PR that enables `NewVolumeManagerReconstruction`, this test should fail.
4. Create a PR that enables `NewVolumeManagerReconstruction` + #117804,  this test should succeed.
5. Merge 4.

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

File: test/e2e/storage/static_pods.go [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+		clusterName = "kind"
+	}
+	return clusterName + "-control-plane"
+}
+
+func prettyCmd(cmd *exec.Cmd) string {
+	return fmt.Sprintf("%s %s", cmd.Path, strings.Join(cmd.Args, " "))
+}
+
+func checkAPIServer(ctx context.Context, expectedAvailable bool, cs clientset.Interface) error {
```
Comment (pohly) [inline test/e2e/storage/static_pods.go:164]: You can drop the `expectedAvailable bool` and return a status string ("running", "stopped", "<some error>"). Then in the `Eventually` calls use `gomega.Equal` to check for the expected state.

I think that makes the test more readable and leads to better failure messages.

File: test/e2e/storage/static_pods.go [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+	// Use a short timeout to check more often.
+	ctx, done := context.WithTimeout(ctx, 5*time.Second)
+	defer done()
```
Comment (pohly) [inline test/e2e/storage/static_pods.go:167]: You could use `Eventually().WithPolling(5 * time.Second)` for this?

File: test/e2e/storage/static_pods.go [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+	_, err := cs.StorageV1().StorageClasses().List(ctx, metav1.ListOptions{})
+	if err != nil {
+		framework.Logf("Connection to the API server failed: %s", err)
+	} else {
+		framework.Logf("Connection to the API server succeeded")
+	}
```
Comment (pohly) [inline test/e2e/storage/static_pods.go:173]: When returning a status, these log messages should no longer be necessary. If the `Eventually` gets stuck, `ginkgo --poll-progress-interval` can be used to be informed while it is running.

File: test/e2e/storage/static_pods.go [hunk 4]
```diff
@@ -0,0 +1,184 @@
...
+		gomega.Eventually(ctx, func(ctx context.Context) error {
+			return checkAPIServer(ctx, false, f.ClientSet)
+		}).WithTimeout(time.Minute).Should(gomega.Succeed())
+
+		ginkgo.By("Stop kubelet")
+		err = stopKindKubelet(ctx)
+		framework.ExpectNoError(err, "Stopping kubelet")
+
+		ginkgo.By("Re-enable control plane")
+		err = enableControlPlane(ctx)
```
Comment (pohly) [inline test/e2e/storage/static_pods.go:102]: There's probably no good way to restore the cluster if this test fails somewhere in the middle?

**Review by pohly — CHANGES_REQUESTED** [review]
I can see how this has to make some assumptions about how KinD operates, but it doesn't seem too bad to me.

File: test/e2e/storage/static_pods.go (hunk 3 above)
Comment (jsafrane) [reply test/e2e/storage/static_pods.go:173]: Logs removed (well, the whole function is now removed)

File: test/e2e/storage/static_pods.go (hunk 2 above)
Comment (jsafrane) [reply test/e2e/storage/static_pods.go:167]: I wanted `List()` below to timeout quickly if the API server is not available, `ctx` would never stop (unless terminated). I can see there is a `metav1.ListOptions{Timeout: *int64}` for that, so I'm using it now.

File: test/e2e/storage/static_pods.go (hunk 1 above)
Comment (jsafrane) [reply test/e2e/storage/static_pods.go:164]: Then I can drop the whole `checkAPIServer` and `List()` something directly in the test. It looks better now.

File: test/e2e/storage/static_pods.go (hunk 4 above)
Comment (jsafrane) [reply test/e2e/storage/static_pods.go:103]: No, I could only delete + re-create the cluster. It's a really disruptive test.

**Review by pohly — COMMENTED** [review]
One nit, otherwise looks good.

File: test/e2e/storage/static_pods.go [hunk 5]
```diff
@@ -0,0 +1,164 @@
...
+			NodeSelection: testConfig.ClientNodeSelection,
+		}
+		pod, err := e2epod.CreateSecPod(ctx, f.ClientSet, podConfig, f.Timeouts.PodStart)
+		framework.ExpectNoError(err, "Creating a pod with a CSI volume")
+		ginkgo.DeferCleanup(func(ctx context.Context) {
+			if pod != nil {
+				err := f.ClientSet.CoreV1().Pods(f.Namespace.Name).Delete(ctx, pod.Name, metav1.DeleteOptions{})
+				framework.ExpectNoError(err, "Deleting the pod with the CSI volume")
+			}
+		})
```
Comment (pohly) [inline test/e2e/storage/static_pods.go:86]: I just noticed one optimization: pod is never going to be nil and while usually `ExpectNoError` is recommended, here the error is going to be simple enough that we can leave error checking to Ginkgo ->

```suggestion
		ginkgo.DeferCleanup(f.ClientSet.CoreV1().Pods(f.Namespace.Name).Delete, pod.Name, metav1.DeleteOptions{})
```

Comment (jsafrane) [reply test/e2e/storage/static_pods.go]: applied

**Comment by aojea** [comment]
we can not have tests like this in the e2e , can this be added in an external repo?

**Comment by aojea** [comment]
for these kind of things we use test-infra experiment, see https://github.com/kubernetes/test-infra/blob/master/experiment/kind-multizone-e2e.sh. it has a strong dependency on the kind cluster but does not make the dependency on the e2e test

**Comment by BenTheElder** [comment]
My concern with this one is that not only does it depend on kind, it depends on kind internals *and* how the test binary and kind cluster are being run, IE it depends on `docker exec` access and `KIND_CLUSTER_NAME` vs `--name`, which is now implied by `Feature:Kind` without being documented (to be fair we introduced comments on the Feature values after this PR).

(I also probably would've called this "Feature:KIND" but /shrug)


When it's done with something like the experiment script, we have the downside of not being able to leverage the e2e framework, but we can explicitly depend on the specifics of the cluster setup.

I think it's neat that we were able to unblock testing this behavior, but I'm concerned about what happens when we have many `Feature:Kind` tests that actually depend on how the cluster is running + kind internals, and when we break those.

node_e2e seems like the closest parallel in-tree, those tests explicitly depend on ssh access to e.g. reboot kubelet, but I'm not sure if they can depend on e.g. the static manifest path.

@kannon92 would this fit for node_e2e? I'm guessing the blocker might be having a CSI driver ...?

**Comment by kannon92** [comment]
Why is this coming up now? The PR merged 2 years ago..

**Comment by kannon92** [comment]
I think that node_e2e could fit but I don't necessary know. We do have tests that stop kubelet in those test suites.

**Comment by BenTheElder** [comment]
> Why is this coming up now? The PR merged 2 years ago..

I just discovered the "feature" while refactoring the e2e framework, there are a handful of still-undocumented "feature"s in the e2e framework.

> I think that node_e2e could fit but I don't necessary know. We do have tests that stop kubelet in those test suites.

ACK

**Pull request merged**